### PR TITLE
Do a pass of typing fixes on transformer/

### PR DESCRIPTION
**Pull request by nschank**
# What does this PR do ?
Simple cleanups of type hints under transformer/

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [X] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [X] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/transformer/moe/token_dispatcher.py [hunk 1]
```diff
@@ -1391,7 +1391,7 @@ def dispatch_preprocess(
     def token_dispatch(
         self,
         hidden_states: torch.Tensor,
-        probs: torch.Tensor = None,
+        probs: Optional[torch.Tensor] = None,
```
Comment (Skylion007) [inline megatron/core/transformer/moe/token_dispatcher.py:1433]: There is a ruff lint you can enable to catch these.

File: megatron/core/transformer/identity_op.py [hunk 2]
```diff
@@ -1,16 +1,24 @@
...
+T = TypeVar('T')
+
 
 class IdentityOp(torch.nn.Module):
     """
     This is a placeholder for IdentityOp(x) -> x
     """
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args: object, **kwargs: object):
```
Comment (Skylion007) [inline megatron/core/transformer/identity_op.py:14]: Weird for this to have kwargs, but okay.

**Review by yashaswikarnati — APPROVED** [review]

**Comment by yashaswikarnati** [comment]
/ok to test [0ac8276](https://github.com/NVIDIA/Megatron-LM/pull/2766/commits/0ac8276d615bd1fc65400582451c3a9b01ead1f3)

File: megatron/core/transformer/identity_op.py (hunk 2 above)
Comment (nschank) [reply megatron/core/transformer/identity_op.py:14]: It's so it can be a drop-in replacement for modules which are constructed with kwargs I assume!

File: megatron/core/transformer/moe/token_dispatcher.py (hunk 1 above)
Comment (nschank) [reply megatron/core/transformer/moe/token_dispatcher.py:1433]: This is a good callout - I can enable that in a followup (it will likely require quite a lot of fixes alongside)

**Comment by nschank** [comment]
I have no idea how, but this was pushed with a missing `]` which was not missing locally - I repushed it and it is now present, sorry for the force push

**Comment by yashaswikarnati** [comment]
/ok to test [7c50e97](https://github.com/NVIDIA/Megatron-LM/pull/2766/commits/7c50e9742b3e45617e036e5a23aab92627dd9e94)

**Comment by nschank** [comment]
As far as I can tell, the failing workflow is not caused by this PR (those things were removed in https://github.com/NVIDIA/Megatron-LM/pull/2474)

**Review by yanring — APPROVED** [review]
LGTM

**Comment by yashaswikarnati** [comment]
/ok to test [7c50e97](https://github.com/NVIDIA/Megatron-LM/pull/2766/commits/7c50e9742b3e45617e036e5a23aab92627dd9e94)
